solidity_contracts/src/CairoPrecompiles/MulticallCairoCounterCaller.sol

```diff
@@ -0,0 +1,209 @@
+        bytes memory result = new bytes(a.length + b.length);
+        uint256 k = 0;
+        for (uint256 i = 0; i < a.length; i++) {
+            result[k] = a[i];
+            k++;
+        }
+        for (uint256 i = 0; i < b.length; i++) {
+            result[k] = b[i];
+            k++;
+        }
+        return result;
+    }
+
+    function batchCallCairo(uint256[][] memory calls) internal returns (bytes memory) {
+        uint32 n_calls = uint32(calls.length);
+        bytes4 selector = bytes4(keccak256("call_contract(bytes4,uint256,uint256,uint256[])"));
+        bytes memory callData = abi.encodePacked(selector, bytes4(n_calls));
+
+        for (uint32 i = 0; i < n_calls; i++) {
+            require(calls[i].length >= 3, "Invalid call format");
+            uint256 contractAddress = calls[i][0];
+            uint256 functionSelector = calls[i][1];
+            uint256 dataLength = calls[i][2];
+            uint256[] memory data = new uint256[](dataLength);
+            for (uint256 j = 0; j < dataLength; j++) {
+                data[j] = calls[i][j + 3];
+            }
+            bytes memory encodedCall = abi.encode(contractAddress, functionSelector, data);
+            callData = concatenateBytes(callData, encodedCall);
+        }
+
+        (bool success, bytes memory result) = BATCH_CAIRO_PRECOMPILE_ADDRESS.call(callData);
+        require(success, string(abi.encodePacked("CairoLib: call_contract failed with: ", result)));
+
+        return result;
+    }
+
+    function callCairo(uint256 contractAddress, uint256 functionSelector, uint256[] memory data)
+        internal
+        returns (bytes memory)
+    {
+        bytes4 selector = bytes4(keccak256("call_contract(bytes4,uint256,uint256,uint256[])"));
+        bytes memory callData =
+            abi.encodePacked(selector, bytes4(uint32(1)), abi.encode(contractAddress, functionSelector, data));
+
+        (bool success, bytes memory result) = BATCH_CAIRO_PRECOMPILE_ADDRESS.call(callData);
+        require(success, string(abi.encodePacked("CairoLib: call_contract failed with: ", result)));
+
+        return result;
+    }
+
+    function callCairo(uint256 contractAddress, uint256 functionSelector) internal returns (bytes memory) {
+        uint256[] memory data = new uint256[](0);
+        return callCairo(contractAddress, functionSelector, data);
+    }
+
+    function callCairo(uint256 contractAddress, string memory functionName, uint256[] memory data)
+        internal
+        returns (bytes memory)
+    {
+        uint256 functionSelector = uint256(keccak256(bytes(functionName))) % 2 ** 250;
+        return callCairo(contractAddress, functionSelector, data);
+    }
+
+    function callCairo(uint256 contractAddress, string memory functionName) internal returns (bytes memory) {
+        uint256[] memory data = new uint256[](0);
+        return callCairo(contractAddress, functionName, data);
+    }
```

Comment:
are those necessary ?
Should it be more restrictive ?
As it increase the code size hence attack surface 

Same for delegate / callcode

Reply:
it's a copy/paste of the CairoLib, for testing purposes only.

Reply:
maybe this should also apply to the cairoLib ?